### PR TITLE
Implement function slice for RecordBatch

### LINKED ISSUE #460: Implement RecordBatch::slice() to slice RecordBatches
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When implementing an operator that needed to check for partitions across `RecordBatch` boundaries (in https://github.com/influxdata/influxdb_iox/pull/1733) I found myself wanting to `slice` the record batch so I could keep part of it while I sent the rest through

In addition I can imagine an operator like `RepartitionExec` which might want to break large batches into smaller batches when repartitioning. 

**Describe the solution you'd like**
Implement the following function

```rust
impl RecordBatch { .
...
  /// Return a new RecordBatch where each column is sliced
  /// according to `offset` and `length`
  pub fn::slice(fn slice(&self, offset: usize, length: usize) -> Result<RecordBatch> {
  }
```

There is already a `Array::slice` method: https://github.com/apache/arrow-rs/blob/master/arrow/src/array/array.rs#L83-L98 so to slice a `RecordBatch` you simply apply that function to each column

Here is a partial implementation, from https://github.com/influxdata/influxdb_iox/pull/1733

```rust
    fn concat_record_batches(batches: &[RecordBatch]) -> ArrowResult<RecordBatch> {
        assert!(!batches.is_empty());

        // concatenate them column by column
        let schema = batches[0].schema();
        let num_columns = batches[0].columns().len();
        let new_columns = (0..num_columns)
            .map(|column_index| {
                let old_columns: Vec<_> = batches
                    .iter()
                    .map(|batch| batch.column(column_index).as_ref())
                    .collect();
                arrow::compute::concat(&old_columns)
            })
            .collect::<ArrowResult<Vec<_>>>()?;

        RecordBatch::try_new(schema, new_columns)
```

**Describe alternatives you've considered**

`RecordBatch::split` is proposed in https://github.com/apache/arrow-rs/issues/343 but slice may be more general (and work better with existing Rust API ecosystem)

**Additional context**

### PR DESCRIPTION
**Pull request by b41sh**
# Which issue does this PR close?

Closes #460

# Rationale for this change
 
 
`slice` can be used to handle part of `RecordBatch`

# What changes are included in this PR?

Implement function `slice` for `RecordBatch`

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/record_batch.rs [hunk 1]
```diff
@@ -244,6 +244,18 @@ impl RecordBatch {
         &self.columns[..]
     }
 
+    /// Return a new RecordBatch where each column is sliced
+    /// according to `offset` and `length`
+    pub fn slice(&self, offset: usize, length: usize) -> Result<RecordBatch> {
+        let schema = self.schema();
+        let num_columns = self.num_columns();
+        let new_columns = (0..num_columns)
```
Comment (jorgecarleitao) [inline arrow/src/record_batch.rs:252]: `self.columns().iter()` ^_^

Comment (nevi-me) [reply arrow/src/record_batch.rs:252]: My local version is

```rust
    /// Slice all the arrays in the record batch.
    pub fn slice(&self, offset: usize, len: usize) -> Self {
        let columns = self
            .columns()
            .iter()
            .map(|array| array.slice(offset, len))
            .collect();
        Self {
            schema: self.schema.clone(),
            columns,
        }
    }
```

Does it make sense to go through `RecordBatch::try_new()`? It incurs some overhead checking that the schema and arrays match, when they already should match

Comment (b41sh) [reply arrow/src/record_batch.rs]: @jorgecarleitao thanks, I have modified, PTAL

Comment (b41sh) [reply arrow/src/record_batch.rs]: @nevi-me Thanks for your suggestion, I made a modification. PTAL

File: arrow/src/record_batch.rs [hunk 2]
```diff
@@ -244,6 +244,21 @@ impl RecordBatch {
         &self.columns[..]
     }
 
+    /// Return a new RecordBatch where each column is sliced
+    /// according to `offset` and `length`
+    pub fn slice(&self, offset: usize, length: usize) -> Result<RecordBatch> {
```
Comment (Jimexist) [inline arrow/src/record_batch.rs:249]: can you remove the `Result` wrapping if nothing can fail.

File: arrow/src/record_batch.rs [hunk 3]
```diff
@@ -426,6 +441,29 @@ mod tests {
         assert_eq!(5, record_batch.column(1).data().len());
     }
 
+    #[test]
+    fn create_record_batch_slice() {
```
Comment (Jimexist) [inline arrow/src/record_batch.rs:445]: can you test for edge cases, e.g. empty length, over boundary offset, empty columns, empty batch, etc.

**Comment by alamb** [comment]
Thank you @b41sh !

File: arrow/src/record_batch.rs (hunk 3 above)
Comment (b41sh) [reply arrow/src/record_batch.rs:445]: @Jimexist I have a question, if an empty RecordBatch calls slice(), the offset and length are zero, should we panic or return an empty RecordBatch slice?

Comment (Dandandan) [reply arrow/src/record_batch.rs:445]: It should return an empty slice.

Comment (b41sh) [reply arrow/src/record_batch.rs:445]: > It should return an empty slice.

ok

Comment (b41sh) [reply arrow/src/record_batch.rs:456]: I have modified, PTAL 
@Dandandan @Jimexist

**Review by alamb — APPROVED** [review]
Thank you @b41sh

**Comment by alamb** [comment]
Hi @b41sh. Thank you for the contribution. 

It appears there are some issues that the CI tests found (clippy): https://github.com/apache/arrow-rs/pull/490/checks?check_run_id=2905850637

Can you please resolve them?

**Comment by b41sh** [comment]
> Hi @b41sh. Thank you for the contribution.
> 
> It appears there are some issues that the CI tests found (clippy): https://github.com/apache/arrow-rs/pull/490/checks?check_run_id=2905850637
> 
> Can you please resolve them?

@alamb thanks for your review, I have fixed the Clippy, PTAL

**Review by alamb — APPROVED** [review]
I think it looks great. Thank you so much @b41sh !

**Review by nevi-me — APPROVED** [review]
